Yoast/wordpress-seo
# Introduce indexation framework

## Context


This introduces a new framework for indexation. Indexation includes tasks such as:
- Internal link counts
- Creating indexables
- Creating prominent words

For internal linking there already exists a new indexation procedure in: https://github.com/Yoast/wordpress-seo-premium/blob/feature/internal-linking-phase-2/premium/src/webservices/post-data-service.php however as this implements the same concepts this framework uses this should be easily adapted to fit the new structure.

This framework primarily revolves around the concept of `Indexation_Action`s. Each action consists of 2-3 parts:
- A PHP function to count the total number of unindexed objects.
- A PHP function to fully index or pre-process a small number unindexed objects.
- Optionally, a JS function to index objects passed from the PHP ( this JS function is assumed to use a normal REST endpoint to save these objects ).

Each indexation action is tied to a rest route, for which the concept of a `Route` and accompanying `Route_Interface` is introduced.

Examples of indexation actions would be:
- An indexable post indexation action
- An indexable term indexation action
- An internal link count indexation action
- etc.

If there is any action with a total unindexed count above 0  then the user needs to index. As each action is tied to a route we can enqueue some JS that will, for each action, call the indexing functions through the REST API and, optionally, registered JS functions until no more unindexed objects remain. As the total count is known beforehand we can render a nice progress bar while doing this.

In order to ensure all actions can be processed exactly the same by the JS driver an `Indexation_Reponse` object is introduced which ensures every indexation rest request has the same response format. The basics of each response are:
- An array of objects that will be passed to the JS indexation if it exists.
- A next url to be called to handle the next batch, this allows pagination params to be passed if required.

This could potentially be expanded with a remaining count but doing so would incur additional queries.

## Summary


This PR can be summarized in the following changelog entry:

* Adds a new tool to the SEO -> Tools page to create indexables for all posts.

## Relevant technical choices:

*

## Test instructions

This PR can be tested by following these steps:

* Run `TRUNCATE TABLE wp_yoast_indexable` to delete all your indexables.
* Go to SEO -> Tools and click on the button to index your content.
* You should have indexables again for every public post.

## UI changes

## Documentation

## Quality assurance
